Scalafix rules to rewrite to new ServerBuilder API automatically

jrudolph opened the pull request:
Still TODO:
 * [x] put them into correct packages
 * [x] some FIXMEs
 * [x] documentation + inclusion in the release notes

ennru commented:
Let's make sure Scala Steward can pick these up so that it can propose the rewrites in its PRs.

jrudolph commented:
> Let's make sure Scala Steward can pick these up so that it can propose the rewrites in its PRs.

I had it on the list but anyway thanks for the reminder, opened PR here: https://github.com/scala-steward-org/scala-steward/pull/1562

jrudolph commented:
Tested the rules on a separate project and added docs section about how to run it in release notes and migration notes.

jrudolph commented:
While testing out required changes for akka-grpc I found one remaining issue:

The new ServerBuilder currently doesn't allow specifying a custom materializer and the scalafix rule will produce broken code. Should we do a last minute fix to add materializer customization to the ServerBuilders and also update the fix rule for those occurrences?

So far, we haven't completely decided if we want to get rid of custom materializers completely or if we want to keep supporting them (which would technically be fine right now).

jrudolph commented:
> The new ServerBuilder currently doesn't allow specifying a custom materializer and the scalafix rule will produce broken code. Should we do a last minute fix to add materializer customization to the ServerBuilders and also update the fix rule for those occurrences?
> 
> So far, we haven't completely decided if we want to get rid of custom materializers completely or if we want to keep supporting them (which would technically be fine right now).

Fixed now in #3414. I now also changed the code to include this case.

jrudolph commented:
Now finished from my side. See 

https://github.com/akka/akka-http/pull/3407/files#diff-d4ea2f20df23bdbfd1e6beedb0020f9aR38-R72

and

https://github.com/akka/akka-http/pull/3407/files#diff-2c37bf7912231bf47cbecc4c9ab209ccR34-R56

for all supported automatic rewrites.

ennru commented on akka-http-scalafix/README.md:
```diff
@@ -0,0 +1,18 @@
+
+ * run `scalafixEnable` on the sbt shell (this will unfortunately require a complete rebuild afterwards)
+ * run `set scalacOptions in ThisBuild += "-P:semanticdb:synthetics:on"` to allow access to synthetics
+ * e.g. run docs/scalafixAll MigrateToServerBuilder`
```
```suggestion
 * e.g. run `docs/scalafixAll MigrateToServerBuilder`
```

ennru commented on akka-http-scalafix/scalafix-rules/src/main/scala/akka/http/fix/MigrateToServerBuilder.scala:
```diff
@@ -0,0 +1,111 @@
+        case (_, matArg) =>
+          Patch.lint(Diagnostic(
+            "custom-materializer-warning",
+            "Custom materializers are often not needed any more. You can often just remove custom materializers and " +
+              "just use the system materializer which is supplied automatically.",
```
```suggestion
            "Custom materializers are often not needed any more. You can often remove custom materializers and " +
              "use the system materializer which is supplied automatically.",
```

ennru commented on akka-http-scalafix/scalafix-rules/src/main/scala/akka/http/fix/MigrateToServerBuilder.scala:
```diff
@@ -0,0 +1,111 @@
+    def handlerIsRoute(handler: Term): Boolean =
+      handler.symbol.info.exists(_.signature.toString contains "Route") || // doesn't seem to work with synthetics on
+        (handler.synthetics match { // only works with `scalacOptions += "-P:semanticdb:synthetics:on"`
+          case ApplyTree(fun, _) :: Nil => fun.symbol.exists(_.displayName == "routeToFlow") /* FIXME: could use real symbol instead */
```
Want to clear the FIXME?

jrudolph replied:
Clarifying the comment instead:

```suggestion
          case ApplyTree(fun, _) :: Nil => fun.symbol.exists(_.displayName == "routeToFlow") // somewhat inaccurate, but that name should be unique enough for our purposes
```

ennru commented on akka-http-scalafix/scalafix-test-input/src/main/scala/akka/http/fix/MigrateToServerBuilderTest.scala:
```diff
@@ -0,0 +1,77 @@
+    def route: Route = ???
+  }
+  def service: ServiceRoutes = ???
+
+  // fix: materializer or system explicitly given, see https://github.com/akka/akka-http/issues/3410
```
Can be removed.

jrudolph replied:
```suggestion
```

jrudolph commented on akka-http-scalafix/scalafix-test-output/src/main/scala/akka/http/fix/MigrateToServerBuilderTest.scala:
```diff
@@ -0,0 +1,61 @@
+    def route: Route = ???
+  }
+  def service: ServiceRoutes = ???
+
+  // fix: materializer or system explicitly given, see https://github.com/akka/akka-http/issues/3410
```
```suggestion
```
